Custom SDP implementation runtime dispatch

drisspg opened the pull request:
# Summary
This shows the run mechanism for dispatching to a custom backend kernel or defaulting to the composite version for SDP. Currently we short circuit for composite but in the future this will dispatch to the appropriate kernel.

drisspg committed:
initial dispatch system stub

drisspg commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -217,36 +211,142 @@ Tensor NestedTensor_to_padded_tensor_cuda(
+
+    // Determine which efficient kernel to use
+    sdp::sdp_params kernel_params{query_, key, value, attn_mask_.has_value(), dropout_p, is_causal};
+    if (sdp::use_flash_attention(kernel_params)){
```
replace with  select_sdp_backend

drisspg commented:
Notes:
- Did not implement context manager versions of use_fused and use_math
- Not tested (YAY!). Not totally true since we have test_transformers that exercised fused_sdp function. However we do not build with flash_attention on CI/CD. As well we don't have A100 that can execute the kernels. SO not totally sure how to do this.

drisspg commented on aten/src/ATen/native/transformers/sdp_utils.h:
```diff
@@ -0,0 +1,252 @@
+      check_packed_projection_shape,
+      check_for_attn_mask,
+      check_cuda_tensors,
+      compiled_with_fused_kernels};
```
Add a cuda arch check?

drisspg replied:
added. Lets see if CI likes this change

cpuhrsch commented on aten/src/ATen/Context.h:
```diff
@@ -126,6 +126,19 @@ class TORCH_API Context {
+  // Note [Disabling Fused SDP Kernels]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // Fused SDP kernels are enabled by default. However, they can be disabled
+  // by setting the at::globalContext().setDisableFusedSDP(true) flag.
```
nit: did you mean to update `setDisableFusedSDP`?

cpuhrsch commented on aten/src/ATen/Context.h:
```diff
@@ -126,6 +126,19 @@ class TORCH_API Context {
+  // by setting the at::globalContext().setDisableFusedSDP(true) flag.
+  // This is useful for debugging purposes. For example, if you want to
+  // compare the performance of the fused SDP kernels with the unfused
+  // kernels, you can disable the fused SDP kernels.
```
You can also force your code to use fused  kernels only if you disable the math kernels.

drisspg replied:
will add

cpuhrsch commented on torch/backends/cuda/__init__.py:
```diff
@@ -160,5 +161,21 @@ def preferred_linalg_library(backend: Union[None, str, torch._C._LinalgBackend]
+
+def set_fused_sdp_enabled(enabled: bool):
+    r""" Enables or disables fused sdp. """
+    torch._C._set_enabled_fused_sdp(enabled)
```
nit: there's two verbs and each in a different tense. Why not `set_sdp_use_fused` or `enable_fused_sdp`?

drisspg commented on torch/csrc/Module.cpp:
```diff
@@ -467,7 +467,36 @@ PyObject* THPModule_float32MatmulPrecision(
+PyObject* THPModule_setUserEnabledFusedSDP(PyObject* _unused, PyObject* arg) {
+  THPUtils_assert(
+      PyBool_Check(arg),
+      "set_enabled_fused_sdp expects a bool, "
```
change this message

drisspg replied:
and the ones below to match the function names

drisspg committed:
Support slice and add tests

drisspg committed:
add chunk test

drisspg committed:
add back in the removed comment

drisspg committed:
explicitly note that backwards are not implemented yet

zou3519 committed:
Add functorch M1 shard (#85565)

functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85565
Approved by: https://github.com/malfet, https://github.com/huydhn

amjames committed:
[Docs] Update sparse Maintainers (#85126)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85126
Approved by: https://github.com/cpuhrsch

amjames committed:
Enable dense x bsc mm/addmm (#85308)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85308
Approved by: https://github.com/pearu

amjames committed:
Enable CSC @ CSC addmm (#85379)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85379
Approved by: https://github.com/pearu, https://github.com/cpuhrsch

drisspg committed:
limit and optimize chunk

drisspg committed:
note on tests

NivekT committed:
[DataLoader] Replacing `traverse` function with `traverse_datapipes` (#85667)

This PR deprecates `traverse` function and replaces it with `traverse_datapipes` instead.

While use `DataLoader`, I realized that it is raising `FutureWarning` even though I am not explicitly using `traverse`. What is happening is that `DataLoader` invokes `traverse(dp, only_datapipe=True)`, and the usage of the keyword causes the `only_datapipe` warning to be raised.

```
/home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/utils/data/graph.py:102: FutureWarning: `only_datapipe` is deprecated from `traverse` function and will be removed after 1.13.
  warnings.warn(msg, FutureWarning)
```

A few things we'd like to do:
1. Deprecate the key word arg `only_datapipe`
2. Change the default behavior from `only_datapipe=False` to `only_datapipe=True` in the future
3. Do not raise a warning when users are using the function correctly

This creates a paradox it is impossible for the users to change their code to match the future default behavior (i.e. call `traverse(dp)` without `only_datapipe`):
  - they cannot do so because the default behavior of `traverse` hasn't changed yet, so they must use `only_datapipe=True`
  - if they use `only_datapipe=True`, eventually the kwarg will go away and cause a runtime error; they also get a `FutureWarning` in the present

IIUC, there doesn't seem to be a way to accomplish those 3 goals without replacing the function with a new one that has a different name; hence, this PR. Let me know if there is a better alternative.

If this looks right, I will send a follow up PR in `TorchData`.

Differential Revision: [D39832183](https://our.internmc.facebook.com/intern/diff/D39832183)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85667
Approved by: https://github.com/ejguan

NivekT committed:
[DataPipe] Fix MapDataPipe spawn lambda test (#85668)

The test in its original form fails and I believe it is because the expected result is incorrect, unless we expect different behaviors between `IterDataPipe` and `MapDataPipe` in multiprocessing.

Differential Revision: [D39832182](https://our.internmc.facebook.com/intern/diff/D39832182)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85668
Approved by: https://github.com/ejguan

fegin committed:
[FSDP] Handle the state_dict on CPU cases (#85640)

state_dict may not be on GPUs. We need to move it to the compute_device in order to gather the ShardedTensor.

Differential Revision: [D39681730](https://our.internmc.facebook.com/intern/diff/D39681730/)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85640
Approved by: https://github.com/rohan-varma, https://github.com/awgu

drisspg committed:
trim

abhudev committed:
[MPS] Enable adaptive avg pool 2d with larger output size (#85726)

* Handle adpative pool 2d forward and backward when ouptut size is larger than input size
* Disallow larger output size if not a multiple of input size
Fixes: https://github.com/pytorch/pytorch/issues/80732
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85726
Approved by: https://github.com/malfet

drisspg committed:
crumbs

malfet committed:
C++20 compatible changes (#85703)

`std::hash<T>::result_type` is deprecated since C++17 and removed in c++20, so use `c10::invoke_result_t` to define it

Fixes https://github.com/pytorch/pytorch/issues/85603

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85703
Approved by: https://github.com/ezyang

jerryzh168 committed:
[quant][docs] Add types for scale and zero_point tensor for torch.fake_quantize_per_channel_affine docs (#85733)

Summary:
Fixes: https://github.com/pytorch/pytorch/issues/85525

Test Plan:
visual inspection for the docs page

Reviewers:

Subscribers:

Tasks:

Tags:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85733
Approved by: https://github.com/z-a-f

andrewor14 committed:
[Quant][fx] Add quant and scale ranges to BackendConfig (#85200)

**Summary:** This commit adds the following constraints to
BackendConfig:

    quant_min_lower_bound
    quant_max_upper_bound
    scale_min_lower_bound
    scale_max_upper_bound

This is motivated by QNNPACK constraints on qint8 weight
values and the min scale value. Actually enforcing these
constraints in the QNNPACK BackendConfig will follow in a
future commit.

Today, users can also specify the above constraints through
QConfigs, and these settings may not necessarily match the
ones specified in the BackendConfig. In this case, we will
handle the discrepancy as follows:

(1) Require QConfig quant ranges to fall within the backend's
(2) Require QConfig min scale value (eps) >= backend's
(3) Require QConfig to specify quant range if the backend
    specified one
(4) Require QConfig to specify min scale value (eps) if the
    backend specified one

Public API changes:

* Previous API, still supported after this commit:
```
dtype_config = DTypeConfig(
    input_dtype=torch.quint8,
    output_dtype=torch.quint8,
    weight_dtype=torch.qint8,
    bias_dtype=torch.float,
)
```
* New API:
```
dtype_config = DTypeConfig(
    input_dtype=DTypeWithConstraints(
        dtype=torch.quint8,
        quant_min_lower_bound=0,
        quant_max_upper_bound=127,
        scale_min_lower_bound=2 ** -12,
    ),
    output_dtype=DTypeWithConstraints(
        dtype=torch.quint8,
        quant_min_lower_bound=0,
        quant_max_upper_bound=127,
        scale_min_lower_bound=2 ** -12,
    ),
    weight_dtype=DTypeWithConstraints(
        dtype=torch.qint8,
        quant_min_lower_bound=-128,
        quant_max_upper_bound=127,
        scale_min_lower_bound=2 ** -12,
    ),
    bias_dtype=torch.float,
)
```
* Additionally, the following `DTypeConfig` attributes
have new types with helper getters:
```
# These have type DTypeWithConstraints
dtype_config.input_dtype
dtype_config.output_dtype
dtype_config.weight_dtype

# These return Optional[torch.dtype]
dtype_config.get_input_dtype()
dtype_config.get_output_dtype()
dtype_config.get_weight_dtype()
```

Note that scale_max is currently not used because there is
no existing mechanism to enforce this on the observer. In the
future, we can validate this as well if there is a use case.

**Test Plan:**

python test/test_quantization.py
TestBackendConfig.test_dtype_with_constraints

python test/test_quantization.py
TestQuantizeFx.test_backend_config_scale_min

python test/test_quantization.py
TestQuantizeFx.test_backend_config_quantization_range

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85200
Approved by: https://github.com/jerryzh168

ezyang committed:
Stop using restore() in ProxyTorchDispatchMode (#85756)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85756
Approved by: https://github.com/samdow

Commit:
[ONNX] Apply Common Subexpression Elimination pass to ONNX export (#85665)

## Summary
Exporting graphs with Autocast may fail due to a limitation on JIT tracer. By disabling Autocast cache, tracer works, but there can be performance hit when there is reuse of weights in convolution, for example

By applying CSE, such performance loss can be reverted.

ps: As a comment at #84092 mentioned, disabling Autocast cache is an acceptable workaround and used throughout PyTorch code.

Fixes #84092

## Examples of before and after CSE being applied:

### Example: eliminating `%17` and reusing `%16` instead

```python
# BEFORE
graph(%0 : Float(requires_grad=0, device=cpu)):
  %3 : Scalar = aten::ScalarImplicit(%0), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %13 : int = prim::Constant[value=3](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %14 : int = prim::Constant[value=4](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %15 : int[] = prim::ListConstruct(%13, %14), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %16 : NoneType = prim::Constant(), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %17 : NoneType = prim::Constant(), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %18 : Device = prim::Constant[value="cpu"](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %19 : bool = prim::Constant[value=0](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %20 : Float(3, 4, strides=[4, 1], requires_grad=0, device=cpu) = aten::full(%15, %3, %16, %17, %18, %19), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  return (%20)

AFTER
graph(%0 : Float(requires_grad=0, device=cpu)):
  %3 : Scalar = aten::ScalarImplicit(%0), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %13 : int = prim::Constant[value=3](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %14 : int = prim::Constant[value=4](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %15 : int[] = prim::ListConstruct(%13, %14), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %16 : NoneType = prim::Constant(), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %18 : Device = prim::Constant[value="cpu"](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %19 : bool = prim::Constant[value=0](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %20 : Float(3, 4, strides=[4, 1], requires_grad=0, device=cpu) = aten::full(%15, %3, %16, %16, %18, %19), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  return (%20)
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85665
Approved by: https://github.com/ngimel, https://github.com/AllenTiTaiWang, https://github.com/BowenBao

H-Huang committed:
[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations (#83810)

### About this PR
* Update the all_reduce op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Update test to validate that a separate device implementation is not supported.

### About this stack
In the future we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D39506979](https://our.internmc.facebook.com/intern/diff/D39506979)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/83810
Approved by: https://github.com/kwen2501

H-Huang committed:
[3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations (#83735)

### About this PR
* Update the broadcast op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Add test to validate that a separate device implementation is not supported.

### About this stack
In the future we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D38876771](https://our.internmc.facebook.com/intern/diff/D38876771)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/83735
Approved by: https://github.com/kwen2501

atalman committed:
Upload to different path for pypi cudnn (#85753)

We want to upload pypi cudnn builds to a different download folder something like cu117_pypi_cudnn

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85753
Approved by: https://github.com/seemethere, https://github.com/kit1980

kumpera committed:
C10D extension to enable per-thread PG (#84153)

 Move a bunch of globals to instance methods and replace all use to them.

We move all PG related globals under World and use a singleton instance under _world.

This creates an undocumented extension point to inject full control of how how c10d
state behaves.

One simple hack is to change _world to an implementation that uses a threadlocal
and enable per-thread PGs.

It almost get DDP working and the PG is missing an implementation of all_reduce.

This enables notebook usage of PTD, which is a big deal for learning it:
https://gist.github.com/kumpera/32cb051fa26b8cad8bdf671f968dcd68

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84153
Approved by: https://github.com/rohan-varma

ezyang committed:
Revert "Revert "Symintifying slice ops (#85196)"" (#85746)

This reverts commit 3a171dfb0c08956d55f341039cf35e3a18269c34.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85746
Approved by: https://github.com/albanD

ezyang committed:
Directly access has_symbolic_sizes_strides, avoid expensive test (#85754)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85754
Approved by: https://github.com/albanD

ezyang committed:
Remove some dead code from fake tensor (#85764)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85764
Approved by: https://github.com/wconstab

slgong-fb committed:
Revisit python tracing OD flow (#85326)

Summary:
- add `set_withstack()`, overriding `ClientInterface`'s no-op funtion.
- revert `start()` and #ifdef

Differential Revision: D39647074

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85326
Approved by: https://github.com/chaekit

wconstab committed:
Auto tag milad for symbolic-shapes PRs (#85751)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85751
Approved by: https://github.com/ezyang

kwen2501 committed:
Add all_gather_into_tensor in place of _all_gather_base (#85686)

### Description
- This PR renames `_all_gather_base` to `all_gather_into_tensor` so that it is clearer in meaning.
- The `all_gather_into_tensor` API differs from the `all_gather` API in the output it accepts -- a single, large tensor instead of a list of tensors.
- This PR also adds deprecation warning to `_all_gather_base`.

### Issue
`_all_gather_base` was implemented in https://github.com/pytorch/pytorch/pull/33924 to avoid unnecessary flattening. There was previous effort (#82639) to merge `_all_gather_base` with the existing `all_gather` API by detecting the parameter type passed in for the output.

There are, however, two "blockers" that make the merge difficult:
(i) The merge leads to backward compatibility break. We would need to change the parameter name `tensor_list` in `all_gather` to a general name `output` that can cover both tensor and tensor list.
(ii) Recently, the `all_gather` API has added uneven tensor support, utilizing the tensor boundaries implied by the list. We are, however, not sure to add such support to the `_all_gather_base` function, because that would require users to pass in additional tensor boundary information.

In view of the above, we decided to productize `_all_gather_base` as a separate function, but with a clearer name.

### Testing
Added tests:
- `test_all_gather_into_cat_tensor_cuda` -- output form as with `torch.cat`. For example:
```
        >>> tensor_in
        tensor([1, 2], device='cuda:0') # Rank 0
        tensor([3, 4], device='cuda:1') # Rank 1
        >>> tensor_out
        tensor([1, 2, 3, 4], device='cuda:0') # Rank 0
        tensor([1, 2, 3, 4], device='cuda:1') # Rank 1
```
- `test_all_gather_into_stack_tensor_cuda` -- output form as with `torch.stack`. For example:
```
        >>> tensor_out2
        tensor([[1, 2],
                [3, 4]], device='cuda:0') # Rank 0
        tensor([[1, 2],
                [3, 4]], device='cuda:1') # Rank 1
```
The output form is determined by the shape of the output tensor passed by the user, no flag used.

Cc @rohan-varma @mrshenli @crcrpar @ptrblck @H-Huang
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85686
Approved by: https://github.com/rohan-varma, https://github.com/crcrpar

Chillee committed:
Added floordiv simplification rule needed for models (#85768)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85768
Approved by: https://github.com/ezyang

Chillee committed:
Rename test file from test_pythonkey to test_aotdispatch (#85769)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85769
Approved by: https://github.com/ezyang

exclowd committed:
52189: refractor unreachable Runtime Error (#85725)

Fixes #52189

Since the PR #52228 had gone cold. I made the requested changes and fixed the linting error.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85725
Approved by: https://github.com/lezcano

drisspg committed:
chunks as view

eastzone committed:
[ucc] Remove internal tracing (#85730)

Summary: Remove internal tracing since this was not upstreamed yet.

Test Plan: All PyTorch test should pass.

Differential Revision: D39853937

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85730
Approved by: https://github.com/kwen2501

drisspg committed:
add some extra checks

Commit:
fix ovrsource pytorch build from D39769513 (#85708)

Test Plan: Tested locally, verifying with CI.

Reviewed By: h-friederich

Differential Revision: D39851831

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85708
Approved by: https://github.com/zou3519

Xia-Weiwen committed:
[UT] Fix random failure of test_qconv_transpose1d by skip using hypothesis (#85463)

TestQuantizedConv.test_qconv_transpose1d fails randomly due to hypothesis (according to @jerryzh168).
This PR fixes it by rewriting the test case without hypothesis. Use fixed parameters and `itertools.product` to generate test cases.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85463
Approved by: https://github.com/jerryzh168

kurtamohler committed:
Update `amax/amin/norm/count_nonzero` signatures with `int[*]? dim` (#83300)

Changes `dim` arg to use `int[*]?` type for the following functions in `native_funcitons.yaml`:
* `amax`
* `amin`
* `norm`
* `frobenius_norm`
* `native_norm`
* `count_nonzero`

Part of #29137

Pull Request resolved: https://github.com/pytorch/pytorch/pull/83300
Approved by: https://github.com/ngimel, https://github.com/albanD, https://github.com/kulinseth

kulinseth committed:
[MPS] Handle 1D weight in linear layer  (#85752)

Fixes https://github.com/pytorch/pytorch/issues/84591

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85752
Approved by: https://github.com/malfet

DenisVieriu97 committed:
[MPS] Add tensor::index_put op (#85672)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85672
Approved by: https://github.com/malfet

pytorchmergebot committed:
Revert "Update `amax/amin/norm/count_nonzero` signatures with `int[*]? dim` (#83300)"

This reverts commit 8c7c7ed3221aeeefb63ef2b7a221a5d8b274cda5.

Reverted https://github.com/pytorch/pytorch/pull/83300 on behalf of https://github.com/huydhn due to The commit pin breaks XLA test somehow

jingxu10 committed:
add itt unit test and docstrings (#84848)

Add unit tests and docstrings corresponding to PR https://github.com/pytorch/pytorch/pull/63289
UT:
1. `test_profiler_emit_itt` in `test/test_autograd.py`. This test is merely intended to catch if emit_itt breaks on construction.
2. Test `torch.profiler.itt` functions in `test/test_itt.py`
3. Only testing that emit_itt runs when `record_shapes` option is enabled in `test/test_profiler.py`.

Docstring:
1. add ITT related info into `docs/source/bottleneck.rst`
4. add `torch.profiler.itt` functions to `docs/source/profiler.rst`
5. add docstring to `torch.profiler.itt` functions in `torch/profiler/itt.py`
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84848
Approved by: https://github.com/malfet

kurtamohler committed:
Update `amax/amin/norm/count_nonzero` signatures with `int[*]? dim` (#83300)

Changes `dim` arg to use `int[*]?` type for the following functions in `native_funcitons.yaml`:
* `amax`
* `amin`
* `norm`
* `frobenius_norm`
* `native_norm`
* `count_nonzero`

Part of #29137

Pull Request resolved: https://github.com/pytorch/pytorch/pull/83300
Approved by: https://github.com/ngimel, https://github.com/albanD, https://github.com/kulinseth

slgong-fb committed:
[Profiler] tracking Optimizer (part 2 of Record Optimizer) (#84920)

Summary:
Part 2 of Record Optimizer param_groups and states (https://github.com/pytorch/pytorch/pull/84063)
- hooking from optimizer step
- PyOptCall Type
- declare data type for collection
- python binding
- simple unit test case

Test Plan: buck run mode/opt //caffe2/test:profiler

Differential Revision: D39402667

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84920
Approved by: https://github.com/robieta

kwen2501 committed:
Add environment parse function that supports default value (#85563)

We use "-2" to represent an unset environment variable.
Now adding a util function to attach default value if environment variable is unset.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85563
Approved by: https://github.com/rohan-varma, https://github.com/H-Huang

psaab committed:
[aarch64] Use the correct binary when cross building //caffe2/torch/csrc/deploy:remove_dt_needed (#85632)

Test Plan: CI

Differential Revision: D39807135

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85632
Approved by: https://github.com/ajtulloch

t-vi committed:
Make Python reference for permute accept varargs (#85460)

Fixes #85452

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85460
Approved by: https://github.com/jjsjann123, https://github.com/mruberry, https://github.com/ngimel

r-barnes committed:
Fix Use obj1.is(obj2) warnings (#85688)

Fixes:
```
#    define PYBIND11_DEPRECATED(reason) [[deprecated(reason)]]
                                          ^
/dev/shm/rbarnes/tempfs/pytorch/torch/csrc/autograd/python_variable.cpp:2603:11: warning: 'operator==' is deprecated: Use obj1.is(obj2) instead [-Wdeprecated-declarations]
  if (out == Py_None) {
          ^
/dev/shm/rbarnes/tempfs/pytorch/cmake/../third_party/pybind11/include/pybind11/detail/../pytypes.h:276:5: note: 'operator==' has been explicitly marked deprecated here
    PYBIND11_DEPRECATED("Use obj1.is(obj2) instead")
    ^
/dev/shm/rbarnes/tempfs/pytorch/cmake/../third_party/pybind11/include/pybind11/detail/common.h:136:43: note: expanded from macro 'PYBIND11_DEPRECATED'
#    define PYBIND11_DEPRECATED(reason) [[deprecated(reason)]]
                                          ^
/dev/shm/rbarnes/tempfs/pytorch/torch/csrc/autograd/python_variable.cpp:2627:11: warning: 'operator==' is deprecated: Use obj1.is(obj2) instead [-Wdeprecated-declarations]
  if (out == Py_None) {
          ^
/dev/shm/rbarnes/tempfs/pytorch/cmake/../third_party/pybind11/include/pybind11/detail/../pytypes.h:276:5: note: 'operator==' has been explicitly marked deprecated here
    PYBIND11_DEPRECATED("Use obj1.is(obj2) instead")
    ^
/dev/shm/rbarnes/tempfs/pytorch/cmake/../third_party/pybind11/include/pybind11/detail/common.h:136:43: note: expanded from macro 'PYBIND11_DEPRECATED'
#    define PYBIND11_DEPRECATED(reason) [[deprecated(reason)]]
                                          ^
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85688
Approved by: https://github.com/albanD, https://github.com/ezyang

Chillee committed:
removed compile cache and static argnums (#85783)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85783
Approved by: https://github.com/wconstab

Chillee committed:
Added symbolic shape testing for AOTAutograd (#85789)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85789
Approved by: https://github.com/ezyang

Chillee committed:
Symintified mmm/addmm derivative formulas (#85794)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85794
Approved by: https://github.com/ezyang

Chillee committed:
Registered _like metas (#85793)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85793
Approved by: https://github.com/ezyang

jeffdaily committed:
skip test TestCompositeComplianceCUDA::test_forward_ad_nn_functional_max_unpool2d_cuda_float32 (#85767)

This test was marked as expected failure, but this test is flaky for ROCm but only because ROCm sometimes gets expected success. The test was only marked expected failure due to non-determinism that was already well-known. See the nearby comments.

https://github.com/pytorch/pytorch/blob/a4c94f0739158d2f7fd27f2be59b77f33027e1c7/torch/testing/_internal/common_methods_invocations.py#L11410-L11421

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85767
Approved by: https://github.com/clee2000

ezyang committed:
Don't make parameters have symbolic shapes (#85809)

Parameters won't change size across iterations of the
training loop, so this is a cost-free optimization that
avoids us having to do symbolic math over parameters.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85809
Approved by: https://github.com/albanD

ezyang committed:
Augment errors raised in fx.Interpreter with Node info (#85810)

We have been using this extra error context in the symbolic-shapes
branch and it is quite useful.  Contributing it upstream here.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85810
Approved by: https://github.com/albanD

janeyx99 committed:
[BE] small typo+lint fixes for einsum/sumproduct_pair (#85709)

Easy review! This PR fixes some typos + lints + clarifies some instructions
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85709
Approved by: https://github.com/soulitzer

zhengwy888 committed:
[NN module] speed up _load_from_state_dict (#85743)

Fixes #61398

The original implementation is very slow when the state_dict.keys() is long. This PR only passes relevant keys to the child module.

existing test passes: `pytest test/test_nn.py -k state_dict`
I couldn't figure out a good way to write a new test for this new behavior. Had a new snippet, but it will be flaky if integrated into the main CI because it's a timing based check.
But I can verify that the test took 30s to run, after this PR it only takes 0.5s.

```python
    def test_load_state_dict_large(self):
        # construct a module with 4 levels of module, 10 linear each, leads to 10k items in the dictionary
        import copy
        import time
        base_module = nn.Linear(1,1)
        model = base_module
        for level in range(4):
           model = nn.Sequential(*[copy.deepcopy(model) for _ in range(10)])
        state_dict = model.state_dict()
        self.assertEqual(len(state_dict), 20000)
        st = time.time()
        model.load_state_dict(state_dict, strict=True)
        strict_load_time = time.time() - st
        # it took 0.5 seconds to
        self.assertLess(strict_load_time, 10)
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85743
Approved by: https://github.com/albanD

alugorey committed:
[ROCm] Retry loop implemented to avoid transient memory leak errors (#82607)

### Description
Added a retry loop to memory leak checker to avoid rare case in which ROCM reports a false positive memory leak.

### Issue
Original issue observed as part of this ticket: https://github.com/pytorch/pytorch/issues/62533

### Testing
- Applied changes and built
- python test/test_cuda.py
- Ensure all tests pass

Pull Request resolved: https://github.com/pytorch/pytorch/pull/82607
Approved by: https://github.com/malfet

pytorchmergebot committed:
Revert "Symintified mmm/addmm derivative formulas (#85794)"

This reverts commit 230edd2515367fcb44cea7c40106ff9f6a712a2a.

Reverted https://github.com/pytorch/pytorch/pull/85794 on behalf of https://github.com/janeyx99 due to Sorry, reverting as this breaks an aot_autograd mac test on functorch https://hud.pytorch.org/pytorch/pytorch/commit/230edd2515367fcb44cea7c40106ff9f6a712a2a

eqy committed:
[CUBLAS][CUDA GRAPHS] (re-re-open of #83461) Explicitly set the workspace for cuBLAS handles (#85447)

Now includes @dagitses 's optimizations and fixes for teardown

CC @ngimel @ptrblck
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85447
Approved by: https://github.com/malfet

drisspg committed:
new flash atten api
